coderd/oauth2provider/tokens.go

```diff
@@ -475,6 +480,34 @@ func refreshTokenGrant(ctx context.Context, db database.Store, app database.OAut
 	}, nil
 }
 
+// validateRedirectURI validates redirect URI according to RFC 6749 and returns the parsed URL if valid
+func validateRedirectURI(p *httpapi.QueryParamParser, redirectURIParam string, registeredRedirectURIs []string) *url.URL {
+	if redirectURIParam == "" {
+		return nil
+	}
+
+	// Parse the redirect URI
+	redirectURL, err := url.Parse(redirectURIParam)
+	if err != nil {
+		p.Errors = append(p.Errors, codersdk.ValidationError{
+			Field:  "redirect_uri",
+			Detail: "must be a valid URL",
+		})
+		return nil
+	}
+
+	// RFC 6749: Exact match against registered redirect URIs
+	if slices.Contains(registeredRedirectURIs, redirectURIParam) {
+		return redirectURL
+	}
```

Comment:
suggestion: add a link to the spec https://datatracker.ietf.org/doc/html/rfc6749#section-3.1.2.3